### PR TITLE
assignment cache sensitive to changes in allocation or variation (ported to 3.x UFC)

### PR DESCRIPTION
**Pull request by leoromanovsky**
## Modification

Same scope as: https://github.com/Eppo-exp/eppo-ios-sdk/pull/33

## Additional changes

If the variation or allocation are not available when logging, practically this is not feasible, however the types ARE `Optional`, then log a default value. This mirrors the implementation in: https://github.com/Eppo-exp/eppo-ios-sdk/pull/33

### CODE REVIEW

File: Sources/eppo/EppoClient.swift [hunk 1]
```diff
@@ -237,16 +237,16 @@ public class EppoClient {
         
         // Optionally log assignment
         if flagEvaluation.doLog {
-            if let allocationKey = flagEvaluation.allocationKey,
-               let variation = flagEvaluation.variation,
```
Comment (leoromanovsky) [inline Sources/eppo/EppoClient.swift:241]: Previously this conditional evaluated to `false` and did not fall through if either `flagEvaluation.allocationKey` or `flagEvaluation.variation` where `.none`. However to mirror the implementation in JS this relaxes the constraints and allows us to log, hypothetically, the sentinel null values.

File: Sources/eppo/dto/EppoValue.swift [hunk 2]
```diff
@@ -187,14 +187,4 @@ public class EppoValue : Decodable, Equatable {
             throw Errors.valueNotSet
         }
     }
-    
-    public func toHashedString() throws -> String {
```
Comment (leoromanovsky) [inline Sources/eppo/dto/EppoValue.swift:191]: not used

File: Sources/eppo/AssignmentCache.swift [hunk 3]
```diff
@@ -9,51 +9,56 @@ public struct AssignmentCacheKey {
     var subjectKey: String
     var flagKey: String
     var allocationKey: String
-    var variationValue: EppoValue
+    var variationKey: String
 }
 
 public class InMemoryAssignmentCache: AssignmentCache {
```
Comment (leoromanovsky) [inline Sources/eppo/AssignmentCache.swift:15]: No changes to this class from https://github.com/Eppo-exp/eppo-ios-sdk/pull/33

File: Sources/eppo/AssignmentCache.swift [hunk 4]
```diff
@@ -9,51 +9,56 @@ public struct AssignmentCacheKey {
...
-    private var cache: [String: String] = [:]
-    
+    internal var cache: [CacheKey: CacheValue] = [:]
+
+    internal struct CacheKey: Hashable {
```
Comment (aarsilv) [inline Sources/eppo/AssignmentCache.swift:18]: fancy!

File: Sources/eppo/AssignmentCache.swift [hunk 5]
```diff
@@ -9,51 +9,56 @@ public struct AssignmentCacheKey {
...
+        
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(subjectKey)
+            hasher.combine(flagKey)
+        }
+    }
+
+    internal struct CacheValue: Equatable {
+        let allocationKey: String
+        let variationKey: String
```
Comment (aarsilv) [inline Sources/eppo/AssignmentCache.swift:30]: Should we be hashing these two things as well? (This was the direction we're going in the JS SDK)

File: Sources/eppo/EppoClient.swift (hunk 1 above)
Comment (aarsilv) [reply Sources/eppo/EppoClient.swift:241]: interesting

File: Sources/eppo/dto/EppoValue.swift (hunk 2 above)
Comment (aarsilv) [reply Sources/eppo/dto/EppoValue.swift:191]: 🪓

File: Tests/eppo/AssignmentCacheTests.swift [hunk 6]
```diff
@@ -15,33 +15,66 @@ final class AssignmentCacheTests: XCTestCase {
...
+        XCTAssertFalse(cache.hasLoggedAssignment(key: key1))
+        XCTAssertTrue(cache.hasLoggedAssignment(key: key2))
+
+        cache.setLastLoggedAssignment(key: key1)
+        XCTAssertTrue(cache.hasLoggedAssignment(key: key1))
+        XCTAssertFalse(cache.hasLoggedAssignment(key: key2))
+
+        cache.setLastLoggedAssignment(key: key2)
+        XCTAssertFalse(cache.hasLoggedAssignment(key: key1))
+        XCTAssertTrue(cache.hasLoggedAssignment(key: key2))
```
Comment (aarsilv) [inline Tests/eppo/AssignmentCacheTests.swift:47]: 🔥

File: Tests/eppo/AssignmentCacheTests.swift [hunk 7]
```diff
@@ -15,33 +15,66 @@ final class AssignmentCacheTests: XCTestCase {
...
-        XCTAssertEqual(cache.get(key: key), value, "Cache should return the value that was set for a key.")
-    }
-    
-    func testHas() {
-        let key = "testKey"
-        XCTAssertFalse(cache.has(key: key), "Cache should return false for a key that has not been set.")
-        cache.set(key: key, value: "testValue")
-        XCTAssertTrue(cache.has(key: key), "Cache should return true for a key that has been set.")
+
+    func testOscillatingAllocation() {
```
Comment (aarsilv) [inline Tests/eppo/AssignmentCacheTests.swift:19]: 🙌

File: Tests/eppo/AssignmentCacheTests.swift [hunk 8]
```diff
@@ -15,33 +15,66 @@ final class AssignmentCacheTests: XCTestCase {
...
+
+        cache.setLastLoggedAssignment(key: key2)
+        XCTAssertFalse(cache.hasLoggedAssignment(key: key1))
+        XCTAssertTrue(cache.hasLoggedAssignment(key: key2))
     }
     
-    func testHasLoggedAssignment() {
-        let assignmentKey = AssignmentCacheKey(
-            subjectKey: "Math", 
+    func testOscillatingVariations() {
```
Comment (aarsilv) [inline Tests/eppo/AssignmentCacheTests.swift:50]: 🙌

**Review by aarsilv — APPROVED** [review]
📈

File: Sources/eppo/AssignmentCache.swift (hunk 5 above)
Comment (leoromanovsky) [reply Sources/eppo/AssignmentCache.swift:30]: Not necessary at this time as this is an in-memory cache; I have some follow-up work for storing to disk. In that case we will implement the `Decodable` interface here and an appropriate handler.

**Pull request merged**